### PR TITLE
[Feature] support key partition exchanger for external table sink

### PR DESCRIPTION
**Pull request by stephen-shelby**
## What type of PR is this：
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #
https://github.com/StarRocks/starrocks/issues/21502

When writing data to an Iceberg partitioned table with an insert statement, values for the same partition column should be written to the same file.  And an iceberg sink operator could include multiple partitioned writers. In order to reduce small files, the partitions processed by each pipeline driver are orthogonal. If I use the current `PartitionExchanger`,  each chunk accepted by an iceberg_sink_operator may contain rows of multiple partitions. So I need to perform a second shuffle.  This result memory copy and lower performance. 

In summary, supporting lake partition exchange is mainly to ensure that the partition column values of each chunk which iceberg_sink_operator accepted are the same.  and it only once shuffle when iceberg table sink writes to multiple partitions.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

File: be/src/exec/pipeline/exchange/local_exchange.cpp [hunk 1]
```diff
@@ -123,6 +123,21 @@ PartitionExchanger::PartitionExchanger(const std::shared_ptr<LocalExchangeMemory
           _part_type(part_type),
           _partition_exprs(partition_expr_ctxs) {}
 
+LakePartitionExchanger::LakePartitionExchanger(const std::shared_ptr<LocalExchangeMemoryManager>& memory_manager,
```
Comment (stephen-shelby) [inline be/src/exec/pipeline/exchange/local_exchange.cpp:126]: `LakePartitionExchanger` might not be a good name

**Comment by DorianZheng** [comment]
please add UT

File: be/src/exec/pipeline/exchange/local_exchange.h [hunk 2]
```diff
@@ -197,6 +197,25 @@ class PartitionExchanger final : public LocalExchanger {
     std::vector<std::unique_ptr<ShufflePartitioner>> _partitioners;
 };
 
+// Exchange the local data for lake partition exchange
+class LakePartitionExchanger final : public LocalExchanger {
```
Comment (DorianZheng) [inline be/src/exec/pipeline/exchange/local_exchange.h:201]: Why not reuse `PartitionExchanger`

File: be/src/exec/pipeline/exchange/local_exchange.cpp [hunk 3]
```diff
@@ -159,6 +172,51 @@ Status PartitionExchanger::accept(const ChunkPtr& chunk, const int32_t sink_driv
...
+        if (partition_row_indexes.find(partition_key) == partition_row_indexes.end()) {
+            partition_row_indexes.emplace(std::move(partition_key), std::make_shared<std::vector<uint32_t>>(1, i));
+        } else {
+            partition_row_indexes.at(partition_key)->emplace_back(i);
+        }
+    }
+
+    for (auto& i : partition_row_indexes) {
+        RowIndexPtr indexes = i.second;
+        std::vector<uint32_t> hash_values(indexes->at(0) + 1, HashUtil::FNV_SEED);
```
Comment (ZiheLiu) [inline be/src/exec/pipeline/exchange/local_exchange.cpp:201]:
```suggestion
    std::vector<uint32_t> hash_values(num_rows, HashUtil::FNV_SEED);
    for (auto& [_, indexes] : partition_row_indexes) {
        hash_values[indexes->at(0)] = HashUtil::FNV_SEED;
```

**Review by ZiheLiu — DISMISSED** [review]

File: be/src/exec/pipeline/exchange/local_exchange.cpp [hunk 4]
```diff
@@ -159,6 +172,52 @@ Status PartitionExchanger::accept(const ChunkPtr& chunk, const int32_t sink_driv
...
+    auto& partitions_columns = _channel_partitions_columns[sink_driver_sequence];
+    for (size_t i = 0; i < partitions_columns.size(); ++i) {
+        ASSIGN_OR_RETURN(partitions_columns[i], _partition_expr_ctxs[i]->evaluate(chunk.get()))
+        DCHECK(partitions_columns[i] != nullptr);
+    }
+
+    Partition2RowIndexes partition_row_indexes;
+    for (int i = 0; i < num_rows; ++i) {
+        auto partition_key = std::make_shared<PartitionKey>(std::make_shared<Columns>(partitions_columns), i);
+        if (partition_row_indexes.find(partition_key) == partition_row_indexes.end()) {
```
Comment (liuyehcf) [inline be/src/exec/pipeline/exchange/local_exchange.cpp:192]: keep the result of find to avoid accessing the same key twice.

File: be/src/exec/pipeline/exchange/local_exchange.cpp [hunk 5]
```diff
@@ -159,6 +172,52 @@ Status PartitionExchanger::accept(const ChunkPtr& chunk, const int32_t sink_driv
...
+            partition_row_indexes.emplace(std::move(partition_key), std::make_shared<std::vector<uint32_t>>(1, i));
+        } else {
+            partition_row_indexes.at(partition_key)->emplace_back(i);
+        }
+    }
+
+    std::vector<uint32_t> hash_values(chunk->num_rows());
+    for (auto& i : partition_row_indexes) {
```
Comment (liuyehcf) [inline be/src/exec/pipeline/exchange/local_exchange.cpp:200]: change `i` to more meaningful name

File: be/src/exec/pipeline/exchange/local_exchange.cpp [hunk 6]
```diff
@@ -159,6 +172,52 @@ Status PartitionExchanger::accept(const ChunkPtr& chunk, const int32_t sink_driv
...
+        RowIndexPtr indexes = i.second;
+        hash_values[indexes->at(0)] = HashUtil::FNV_SEED;
```
Comment (liuyehcf) [inline be/src/exec/pipeline/exchange/local_exchange.cpp:202]: don't use at, it has overhead.

File: be/src/exec/pipeline/exchange/local_exchange.cpp [hunk 7]
```diff
@@ -159,6 +172,52 @@ Status PartitionExchanger::accept(const ChunkPtr& chunk, const int32_t sink_driv
...
+        for (const ColumnPtr& column : partitions_columns) {
+            column->fnv_hash(&hash_values[0], indexes->at(0), indexes->at(0) + 1);
+        }
+
+        uint32_t shuffle_channel_id = hash_values[indexes->at(0)] % source_op_cnt;
+
+        size_t memory_usage(0);
```
Comment (liuyehcf) [inline be/src/exec/pipeline/exchange/local_exchange.cpp:209]: memory_usage = 0;

File: be/src/exec/pipeline/exchange/local_exchange_source_operator.cpp [hunk 8]
```diff
@@ -50,22 +50,63 @@ Status LocalExchangeSourceOperator::add_chunk(ChunkPtr chunk, std::shared_ptr<st
...
+        Columns one_row_partitions_columns(partition_expr_ctxs.size());
+        for (size_t i = 0; i < partition_expr_ctxs.size(); ++i) {
+            ASSIGN_OR_RETURN(one_row_partitions_columns[i], partition_expr_ctxs[i]->evaluate(one_row_chunk.get()));
+            DCHECK(one_row_partitions_columns[i] != nullptr);
+        }
+
+        auto copied_partition_key =
+                std::make_shared<PartitionKey>(std::make_shared<Columns>(one_row_partitions_columns), 0);
+
+        auto queue = std::make_shared<std::queue<PartitionChunk>>();
```
Comment (liuyehcf) [inline be/src/exec/pipeline/exchange/local_exchange_source_operator.cpp:76]: seems ptr is unnecessary.

File: be/src/exec/pipeline/exchange/local_exchange_source_operator.cpp [hunk 9]
```diff
@@ -84,8 +127,10 @@ Status LocalExchangeSourceOperator::set_finished(RuntimeState* state) {
 
 StatusOr<ChunkPtr> LocalExchangeSourceOperator::pull_chunk(RuntimeState* state) {
     ChunkPtr chunk = _pull_passthrough_chunk(state);
-    if (chunk == nullptr) {
+    if (chunk == nullptr && _key_partition_pending_chunk_empty()) {
         chunk = _pull_shuffle_chunk(state);
+    } else if (chunk == nullptr && !_key_partition_pending_chunk_empty()) {
+        chunk = _pull_key_partition_chunk(state);
     }
     return std::move(chunk);
```
Comment (liuyehcf) [inline be/src/exec/pipeline/exchange/local_exchange_source_operator.cpp:135]: this logic is not intuitive, can we just set a type to mark it's type? shuffle/passthrough/key_partition? So all the logic of `is_finished` and `has_output` can be simplified.

File: be/src/exec/pipeline/exchange/local_exchange_source_operator.cpp [hunk 10]
```diff
@@ -139,4 +184,61 @@ ChunkPtr LocalExchangeSourceOperator::_pull_shuffle_chunk(RuntimeState* state) {
...
+        chunk->append_selective(*partition_chunk.chunk, partition_chunk.indexes->data(), partition_chunk.from,
+                                partition_chunk.size);
+    }
+
+    return chunk;
+}
+
+int64_t LocalExchangeSourceOperator::_key_partition_max_rows() const {
+    int64_t max_rows = 0;
+    if (!_partitions.empty()) {
```
Comment (liuyehcf) [inline be/src/exec/pipeline/exchange/local_exchange_source_operator.cpp:227]: redundant check.

File: be/src/exec/pipeline/exchange/local_exchange_source_operator.cpp [hunk 11]
```diff
@@ -50,22 +50,63 @@ Status LocalExchangeSourceOperator::add_chunk(ChunkPtr chunk, std::shared_ptr<st
...
+                                              uint32_t from, uint32_t size, Columns& partition_columns,
+                                              const std::vector<ExprContext*>& partition_expr_ctxs,
+                                              size_t memory_usage) {
+    std::lock_guard<std::mutex> l(_chunk_lock);
+    if (_is_finished) {
+        return Status::OK();
+    }
+
+    auto partition_key = std::make_shared<PartitionKey>(std::make_shared<Columns>(partition_columns), indexes->at(0));
+    if (_partitions.find(partition_key) == _partitions.end()) {
```
Comment (liuyehcf) [inline be/src/exec/pipeline/exchange/local_exchange_source_operator.cpp:63]: keep the result of find to avoid accessing the existed key twice.

**Review by liuyehcf — CHANGES_REQUESTED** [review]

File: be/src/exec/pipeline/exchange/local_exchange.h [hunk 12]
```diff
@@ -197,6 +197,26 @@ class PartitionExchanger final : public LocalExchanger {
     std::vector<std::unique_ptr<ShufflePartitioner>> _partitioners;
 };
 
+// key partition mainly means that the column value of each partition is the same.
+// For external table sinks, the chunk received by operators after exchange need to ensure that
+// the values of the partition columns are the same.
+class KeyPartitionExchanger final : public LocalExchanger {
+    using RowIndexPtr = std::shared_ptr<std::vector<uint32_t>>;
```
Comment (Youngwb) [inline be/src/exec/pipeline/exchange/local_exchange.h:204]: What’s the RowIndexPtr meaning?

File: be/src/exec/pipeline/exchange/local_exchange.cpp (hunk 4 above)
Comment (stephen-shelby) [reply be/src/exec/pipeline/exchange/local_exchange.cpp]: done

Comment (stephen-shelby) [reply be/src/exec/pipeline/exchange/local_exchange.cpp]: done

File: be/src/exec/pipeline/exchange/local_exchange_source_operator.h [hunk 13]
```diff
@@ -100,6 +162,7 @@ class LocalExchangeSourceOperator final : public SourceOperator {
     // TODO(KKS): make it lock free
     mutable std::mutex _chunk_lock;
     const std::shared_ptr<LocalExchangeMemoryManager>& _memory_manager;
+    std::map<PartitionKeyPtr, PendingPartitionChunks, PartitionKeyComparator> _partitions;
```
[the remark this answers is not in the document]
Comment (stephen-shelby) [reply be/src/exec/pipeline/exchange/local_exchange_source_operator.h:164]: will optimize it later

File: be/src/exec/pipeline/exchange/local_exchange.h (hunk 12 above)
Comment (stephen-shelby) [reply be/src/exec/pipeline/exchange/local_exchange.h:204]: row index for each partition

**Review by liuyehcf — DISMISSED** [review]

File: be/src/exec/pipeline/exchange/local_exchange.cpp [hunk 14]
```diff
@@ -159,6 +172,54 @@ Status PartitionExchanger::accept(const ChunkPtr& chunk, const int32_t sink_driv
...
+        if (partition_row_index == partition_row_indexes.end()) {
+            partition_row_indexes.emplace(std::move(partition_key), std::make_shared<std::vector<uint32_t>>(1, i));
+        } else {
+            partition_row_index->second->emplace_back(i);
+        }
+    }
+
+    std::vector<uint32_t> hash_values(chunk->num_rows());
+    for (auto& partition_row_index : partition_row_indexes) {
+        RowIndexPtr indexes = partition_row_index.second;
```
Comment (ZiheLiu) [inline be/src/exec/pipeline/exchange/local_exchange.cpp:203]: Could use structured binding declaration `[_, indexes]` to avoid copying indexes and simplification.

```suggestion
    for (auto& [_, indexes] : partition_row_indexes) {
```

File: be/src/exec/pipeline/exchange/local_exchange.cpp [hunk 15]
```diff
@@ -159,6 +172,54 @@ Status PartitionExchanger::accept(const ChunkPtr& chunk, const int32_t sink_driv
...
+        hash_values[(*indexes)[0]] = HashUtil::FNV_SEED;
+        for (const ColumnPtr& column : partitions_columns) {
+            column->fnv_hash(&hash_values[0], (*indexes)[0], (*indexes)[1]);
```
Comment (ZiheLiu) [inline be/src/exec/pipeline/exchange/local_exchange.cpp:206]: The third parameter should be `(*indexes)[0] + 1`?

**Review by Youngwb — APPROVED** [review]

**Review by ZiheLiu — APPROVED** [review]
